components/tabs/tab-autocollapse-titles.js

```diff
@@ -0,0 +1,541 @@
+
+  React.useEffect(
+    () => {
+      const handler = setTimeout(() => {
+        if (value !== debouncedValue) {
+          setDebouncedValue(value);
+        }
+      }, delay);
+
+      return () => {
+        clearTimeout(handler);
+      };
+    },
+    [value, delay, debouncedValue]
+  );
+
+  return debouncedValue;
+}
+
+function visibilityReducer(state, action) {
+  switch (action.type) {
+    case ACTIONS.MEASURE_TABS: {
+      const {more, tabs} = action;
+      return {
+        ...state,
+        sizes: {
+          more,
+          tabs
+        }
+      };
+    }
+    case ACTIONS.DEFINE_LAST_VISIBLE_INDEX: {
+      const {lastVisibleIndex} = action;
+      return {
+        ...state,
+        lastVisibleIndex
+      };
+    }
+    default:
+      throw new Error();
+  }
+}
+
+const useAdjustHandler = ({
+  measureRef, elements, children, selectedIndex, dispatch
+}) => React.useCallback(entry => {
+  const containerWidth = entry ? entry.contentRect.width : measureRef.current.offsetWidth;
+
+  const {tabs: tabsSizes, more = 0} = elements.sizes;
+
+  let renderMore = children.some(tab => tab.props.alwaysHidden);
+
+  const tabsToRender = [];
+  let filledWidth = renderMore ? more ?? 0 : 0;
+
+  for (let i = 0; i < tabsSizes.length; i++) {
+    if (filledWidth + tabsSizes[i] < containerWidth + MEASURE_TOLERANCE) {
+      filledWidth += tabsSizes[i];
+      tabsToRender.push(tabsSizes[i]);
+    } else {
+      break;
+    }
+  }
+
+  if (tabsToRender.length < tabsSizes.length && !renderMore) {
+    for (let i = tabsToRender.length - 1; i >= 0; i--) {
+      if (filledWidth + more < containerWidth + MEASURE_TOLERANCE) {
+        filledWidth += more;
+        renderMore = true;
+        break;
+      } else {
+        filledWidth -= tabsToRender[i];
+        tabsToRender.pop();
+      }
+    }
+  }
+
+  if (selectedIndex > tabsToRender.length - 1) {
+    const selectedWidth = tabsSizes[selectedIndex];
+    for (let i = tabsToRender.length - 1; i >= 0; i--) {
+      if (filledWidth + selectedWidth < containerWidth + MEASURE_TOLERANCE) {
+        filledWidth += selectedWidth;
+        break;
+      } else {
+        filledWidth -= tabsToRender[i];
+        tabsToRender.pop();
+      }
+    }
+  }
+
+  if (elements.lastVisibleIndex !== tabsToRender.length - 1) {
+    dispatch({
+      type: ACTIONS.DEFINE_LAST_VISIBLE_INDEX,
+      lastVisibleIndex: tabsToRender.length - 1
+    });
+  }
+
+}, [children, dispatch, elements.lastVisibleIndex, elements.sizes, measureRef, selectedIndex]);
+
+export const TabAutocollapseTitles = ({
+  children,
+  selected,
+  onSelect,
+  moreClassName,
+  moreActiveClassName,
+  morePopupClassName
+}) => {
+  const [elements, dispatch] = React.useReducer(visibilityReducer, DEFAULT_STATE);
+  const [preparedElements, setPreparedElements] = React.useState({visible: [], hidden: []});
+
+  const elementsToRender = useDebounce(preparedElements, DEFAULT_DEBOUNCE_INTERVAL);
```

Comment:
I think it’s better to debounce the effect function that calls `setPreparedElements` (https://github.com/JetBrains/ring-ui/blob/ae126037246695f89c3e074e5b30d716502da0a0/components/tabs/tab-autocollapse-titles.js#L379-L406)